stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -112,19 +114,30 @@ internal class ChannelControllerImpl(
     val unfilteredMessages = _messages.map { it.values.toList() }
 
     /** a list of messages sorted by message.createdAt */
-    override val messages: LiveData<List<Message>> = messagesTransformation(_messages)
+    val sortedVisibleMessages = messagesTransformation(_messages).stateIn(domainImpl.scope, SharingStarted.Eagerly, emptyList())
+    override val messages: LiveData<List<Message>> = sortedVisibleMessages.asLiveData()
+
+    private val _messagesState: StateFlow<ChannelController.MessagesState> =
+        _loading.combine(sortedVisibleMessages) { loading: Boolean, messages: List<Message> ->
+            when {
+                loading -> ChannelController.MessagesState.Loading
+                messages.isEmpty() -> ChannelController.MessagesState.OfflineNoResults
+                else -> ChannelController.MessagesState.Result(messages)
+            }
+        }.stateIn(domainImpl.scope, SharingStarted.Eagerly, ChannelController.MessagesState.NoQueryActive)
+    override val messagesState = _messagesState.asLiveData()
 
-    override val oldMessages: LiveData<List<Message>> = messagesTransformation(_oldMessages)
+    override val oldMessages: LiveData<List<Message>> = messagesTransformation(_oldMessages).asLiveData()
 
-    private fun messagesTransformation(messages: MutableStateFlow<Map<String, Message>>): LiveData<List<Message>> {
+    private fun messagesTransformation(messages: MutableStateFlow<Map<String, Message>>): Flow<List<Message>> {
```

Comment:
Nit: with flows, it's convenient to use such things as extensions. It let you chain operators in a row